### ISSUE TITLE
need to sanitize some html codes in the edit window

### ISSUE THREAD

**Issue by rgetz** [splitbrain/dokuwiki#3366]
I recently noticed that when copying in some things from other web sites, the code <U+200B> (Unicode Character ['ZERO WIDTH SPACE')](https://www.fileformat.info/info/unicode/char/200b/index.htm) will sometimes creep in places it shouldn't. (and it's displayed in the edit window as a zero width space, so it's nearly impossible to find)

For example, where I noticed it (and didn't fix it yet) was on a docuwiki [here](https://wiki.analog.com/resources/tools-software/uc-drivers/renesas/ad7980)

has a link it in, that looks like: `[[adi>​PulsarPMODs|EVAL-AD7980-PMDZ]]`

however, on the disk, if you look at the .txt file, it's actually: `[[adi><U+200B>PulsarPMODs|EVAL-AD7980-PMDZ]]`

and you get a url, that has the character encoded (and isn't what the person expects) `https://www.analog.com/%E2%80%8BPulsarPMODs`

While I can fix it from this specific page - a general purpose (I don't know what the "right" fix is - likely to remove it in the text window, and/or to make it a specific tag <zwsp> or something else) solution would be better, so it doesn't happen to others, or me again.

**Comment by splitbrain** [comment]
I think this is mainly a problem with interwiki or external links. Internal links will be cleaned from special chars, so a zerowidth whitespace wouldn't matter. I don't think we can simply apply `stripspecials` for for interwiki links since we know nothing abut the capabilities of the external system.

I think the best solution would be to make special chars visible in the editor so users can see and fix these mistakes. I'm not sure how, yet. I think it should be possible to create a custom font containing visible glyphs for a set of invisible characters, that font could then be used as the first font for the text area, falling back to the second font for all undefined, visible characters...

**Comment by rgetz** [comment]
I have seen similar (potentially related) when end users copy text from a `<code>` or `<term>` block - and paste into their terminal - it doesn't function due to some invisible characters. I will see if I can reproduce this - and investigate the root cause of failure...

**Comment by rgetz** [comment]
I did find:
- https://www.soscisurvey.de/tools/view-chars.php
- https://www.textmagic.com/free-tools/unicode-detector

looks like I can replicate things in github editor as well. (so you don't need to go anywhere else to see it):
[link](http://www.analog.com/​PulsarPMODs) `[link](http://www.analog.com/​PulsarPMODs)`

I will reach out to github support and see what they say as well...

**Comment by rgetz** [comment]
Answer from github support:

>Thank you for sharing your thoughts about those pesky zero width characters us!
Your ticket was automatically routed to our Developer Support team but our Product team is better suited to help you out. We have forwarded your ticket to the Product team who reads and evaluates all feedback and feature requests, however we cannot guarantee a response to every submission.
Please note that this means the Developer Support team will no longer be able to see changes to the ticket, so if you have a follow up that requires further support, please create a new ticket. Thank you again for your feedback!

So, I read that as "if it hurts when you do that, don't do that". (developer support doesn't consider it a "bug").

**Comment by phy25** [comment]
Interesting issue... I came across https://www.fileformat.info/info/unicode/char/200B/fontsupport.htm which listed 3 fonts that shows the zero width space differently (which are actually the 3 fonts in https://en.wikipedia.org/wiki/Fallback_font):

- LastResort
- Unicode BMP Fallback SIL
- Unifont

Unfortunately I didn't find any font that only contains "default-ignorable characters" (http://unicode.org/faq/unsup_char.html). We can probably make one but I don't know if there are browser implementations that will prevent these unicode characters from being rendered (which is probably when we need a rich text editor...)

**Comment by micgro42** [comment]
Yes that is a notorious problem. It is so annoying that I even developed a plugin against it: https://www.dokuwiki.org/plugin:preventzerowidthchars

Related issues and PRs:
* #1208 
* #2405
* #2618 (resolving the issue with zero-width-chars in our own diffs, but they can still be copied from somewhere else)
